proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,497 @@
+  fine; GHC would discover that ``_`` stands for ``Bool``. This is *not* universal quantification.
+  The elision is just a spot about which we don't care, and want GHC to do its best.
+  In ``g ::? forall _. _ -> _; g x = x``, we actually generalize over the elision, so it's not
+  exactly existential quantification. (Without the ``forall _``, it would be.)
+
+Unresolved Questions
+--------------------
+* Is this really the best syntax? I am uncomfortable at stealing both underscored-identifiers and ``::?``.
+  How painful is it to do so?
+
+* Is this design too elaborate? I have a tendency to build elaborate but expressive edifices. Perhaps
+  there is a sweet spot closer to the ground here.
+  
+Further examples
+----------------
+
+.. _below:
+  
+
+* Let's dive deeper into this example::
+
+    silly ::? Proxy a -> ()
+    silly (_ :: Proxy @Bool _) = ()
+
+  Actually, let's first consider something very closely related::
+
+    sillier :: Proxy a -> ()
+    sillier (_ :: Proxy @Bool _) = ()
+
+  This definition of ``sillier`` is rejected. That's because GHC processes its type signature in isolation.
+  GHC sees that we wish to quantify over ``a``. After kind-checking, the kind of ``a`` is utterly unconstrained.
+  Thus, GHC infers ``sillier :: forall {k} (a :: k). Proxy @k a -> ()``. All of this has happened *without*
+  looking at the definition of ``sillier``. When GHC does look at that definition, it is rejected, as
+  the argument to ``sillier`` has type ``Proxy @k a``, not ``Proxy @Bool a``. This is the kind-level
+  equivalent of something like ::
+
+    silliest :: a -> a
+    silliest True = False
+
+  In both ``sillier`` and ``silliest``, the type signature is *more general* than the definition. It's
+  just that ``sillier`` is harder to see that.
+
+  Returning to ``silly``, with the partial type signature marker ``::?``, GHC will *not* kind-generalize
+  the type. It will effectively infer ``silly ::? forall (a :: _k). Proxy @_k a -> ()``, where ``_k``
+  behaves like a named wildcard. (I say "behaves like" because named wildcards become unification variables
+  internally; in this case, the kind of ``a`` really would just be a unification variable.) Now, when
+  checking the definition of ``silly``, GHC is free to discover that ``_k`` should be ``Bool``, and all is
+  well.
+
+  Note that I did *not* mean to write ``silly ::? Proxy _a -> ()``. I want ``a`` to be a skolem here. It's
+  ``a``\s *kind* that I want not to be a skolem.
+
+  Thus, ``silly`` is silly only because it is contrived, not because it is wrong.
+
+* Here are some examples around the use of elisions:
+
+  1. ::
+
+       f :: _ -> _
+
+     This would mean the same as ``f :: a -> b``.
+
+  2. ::
+
+       f :: forall a. a -> _
+
+     This would mean the same as ``f :: forall a. forall b. a -> b`` (the second ``forall`` prevents
+     ``b`` from being in scope in the defined term). I'm sure you now ask: "Why the unrequested
```

Comment:
Can a type elision introduce constraints as well? e.g. how would we treat

```
f :: forall a. a -> _ -> _
f _ x = x + 1
```

Reply:
In this example, we have a complete type signature. That means the signature is fully understood without looking at the term. It thus becomes equivalent to `forall a. forall b c. a -> b -> c`. The definition is then rejected.

If we had a partial type signature (with `::?`), then the definition is still rejected, because there is no extra-constraints wildcard.

Reply:
It seems a bit odd that type elision can introduce new quantifiers but not new constraints. What's the underlying principle? It can't be avoiding spooky action at a distance because we already have that with quantifiers. Perhaps we want to avoid introducing new constraints because they have a runtime cost. I don't find this argument all that compelling either since the user opted in by writing the elision.

But regardless, the proposal should specify and motivate this behavior.